In [125]:
import pandas as pd

# Recruitment Data dataset

In [126]:
# Reading the dataset

In [127]:
df = pd.read_csv("data/recruitment_data.csv")

In [128]:
df

Age  Gender  EducationLevel  ExperienceYears  PreviousCompanies  \
0      26       1               2                0                  3   
1      39       1               4               12                  3   
2      48       0               2                3                  2   
3      34       1               2                5                  2   
4      30       0               1                6                  1   
...   ...     ...             ...              ...                ...   
1495   48       0               2                3                  4   
1496   27       1               2               10                  3   
1497   24       1               1                1                  2   
1498   48       0               2                4                  4   
1499   34       1               2               11                  5   

      DistanceFromCompany  InterviewScore  SkillScore  PersonalityScore  \
0               26.783828              48          78                91   
1               25.862694              35          68                80   
2                9.920805              20          67                13   
3                6.407751              36          27                70   
4               43.105343              23          52                85   
...                   ...             ...         ...               ...   
1495             9.183783              66           3                80   
1496            14.847731              43          97                 7   
1497             4.289911              31          91                58   
1498            36.299263               9          37                44   
1499            12.910472              63          40                26   

      RecruitmentStrategy  HiringDecision  
0                       1               1  
1                       2               1  
2                       2               0  
3                       3               0  
4                       2               0  
...                   ...             ...  
1495                    3               1  
1496                    2               0  
1497                    1               1  
1498                    2               1  
1499                    2               1  

[1500 rows x 11 columns]

In [129]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1500 entries, 0 to 1499
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Age                  1500 non-null   int64  
 1   Gender               1500 non-null   int64  
 2   EducationLevel       1500 non-null   int64  
 3   ExperienceYears      1500 non-null   int64  
 4   PreviousCompanies    1500 non-null   int64  
 5   DistanceFromCompany  1500 non-null   float64
 6   InterviewScore       1500 non-null   int64  
 7   SkillScore           1500 non-null   int64  
 8   PersonalityScore     1500 non-null   int64  
 9   RecruitmentStrategy  1500 non-null   int64  
 10  HiringDecision       1500 non-null   int64  
dtypes: float64(1), int64(10)
memory usage: 129.0 KB


It looks like the dataset has no missing values and all fields data types make sense. Some of the variables could be converted to categorical but I will leave them for now. 

In [130]:
df.describe()

Age       Gender  EducationLevel  ExperienceYears  \
count  1500.000000  1500.000000     1500.000000      1500.000000   
mean     35.148667     0.492000        2.188000         7.694000   
std       9.252728     0.500103        0.862449         4.641414   
min      20.000000     0.000000        1.000000         0.000000   
25%      27.000000     0.000000        2.000000         4.000000   
50%      35.000000     0.000000        2.000000         8.000000   
75%      43.000000     1.000000        3.000000        12.000000   
max      50.000000     1.000000        4.000000        15.000000   

       PreviousCompanies  DistanceFromCompany  InterviewScore   SkillScore  \
count         1500.00000          1500.000000     1500.000000  1500.000000   
mean             3.00200            25.505379       50.564000    51.116000   
std              1.41067            14.567151       28.626215    29.353563   
min              1.00000             1.031376        0.000000     0.000000   
25%              2.00000            12.838851       25.000000    25.750000   
50%              3.00000            25.502239       52.000000    53.000000   
75%              4.00000            37.737996       75.000000    76.000000   
max              5.00000            50.992462      100.000000   100.000000   

       PersonalityScore  RecruitmentStrategy  HiringDecision  
count       1500.000000          1500.000000     1500.000000  
mean          49.387333             1.893333        0.310000  
std           29.353201             0.689642        0.462647  
min            0.000000             1.000000        0.000000  
25%           23.000000             1.000000        0.000000  
50%           49.000000             2.000000        0.000000  
75%           76.000000             2.000000        1.000000  
max          100.000000             3.000000        1.000000

Distribution of key features looks reasonable (Gender, Age, Scores, Experience, etc.)

The column names are currently CamelCase. I will convert them to snake_case.

In [131]:
df.columns = df.columns.str.strip().str.replace(r'(?<!^)(?=[A-Z])', '_', regex = True).str.lower()

I will drop unnecessary columns.

In [132]:
df = df.drop(columns = "distance_from_company")
df = df.drop(columns = "previous_companies")
df = df.drop(columns = "recruitment_strategy")

I will convert the gender, eduaction_level, and hiring_decision columns to categorical variables and replace the digits with the corresponding values.

In [133]:
df['gender'] = df['gender'].replace({0: 'Male', 1: 'Female'})
df['education_level'] = df['education_level'].replace({
    1: "Bachelor's",
    2: "Bachelor's",
    3: "Master's",
    4: "PhD"
})
df['hiring_decision'] = df['hiring_decision'].replace({0: 'Not Hired', 1: "Hired"})

In [134]:
df['gender'] = df['gender'].astype('category')
df['education_level'] = df['education_level'].astype('category')
df['hiring_decision'] = df['hiring_decision'].astype('category')

In [135]:
df.to_csv('data/cleaned_recruitment_data.csv', index = False)

## Dataset Metadata – cleaned_recruitment_data

**File Name**: `cleaned_recruitment_data.csv`  
**Records**: 1,500  
**Features**: 8  
**Target Variable**: `hiring_decision` (Not Hired, Hired) 


| Feature | Description | Type | Range / Categories |
|--------|-------------|------|---------------------|
| `age` | Age of the candidate | Integer | 20–50 |
| `gender` | Candidate's gender | Categorical | 0: Male, 1: Female |
| `education_level` | Highest education attained | Categorical (1–4) | 1: Bachelor's, 2: Master's, 3: PhD |
| `experience_years` | Years of experience | Integer | 0–15 |
| `interview_score` | Score from the interview | Integer | 0–100 |
| `skill_score` | Score on technical assessment | Integer | 0–100 |
| `personality_score` | Personality traits evaluation | Integer | 0–100 |
| `hiring_decision` | Final hiring outcome | Categorical | 0: Not Hired, 1: Hired |

# Resume Dataset

In [136]:
# Reading the dataset

In [137]:
df_r = pd.read_csv('data/UpdatedResumeDataSet.csv')

In [138]:
df_r

Category                                             Resume
0    Data Science  Skills * Programming Languages: Python (pandas...
1    Data Science  Education Details \r\nMay 2013 to May 2017 B.E...
2    Data Science  Areas of Interest Deep Learning, Control Syste...
3    Data Science  Skills • R • Python • SAP HANA • Table...
4    Data Science  Education Details \r\n MCA   YMCAUST,  Faridab...
..            ...                                                ...
957       Testing  Computer Skills: • Proficient in MS office (...
958       Testing  ❖ Willingness to accept the challenges. ❖ ...
959       Testing  PERSONAL SKILLS • Quick learner, • Eagerne...
960       Testing  COMPUTER SKILLS & SOFTWARE KNOWLEDGE MS-Power ...
961       Testing  Skill Set OS Windows XP/7/8/8.1/10 Database MY...

[962 rows x 2 columns]

In [139]:
df_r.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 962 entries, 0 to 961
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Category  962 non-null    object
 1   Resume    962 non-null    object
dtypes: object(2)
memory usage: 15.2+ KB


In [140]:
df_r.describe()

Category                                             Resume
count              962                                                962
unique              25                                                166
top     Java Developer  Technical Skills Web Technologies: Angular JS,...
freq                84                                                 18

We have 25 job categories and 962 resumes. Apparently, there are duplicate resumes, given that we have 166 unique ones.

In [141]:
df_r[df_r.duplicated() == True]

Category                                             Resume
10   Data Science  Skills * Programming Languages: Python (pandas...
11   Data Science  Education Details \r\nMay 2013 to May 2017 B.E...
12   Data Science  Areas of Interest Deep Learning, Control Syste...
13   Data Science  Skills • R • Python • SAP HANA • Table...
14   Data Science  Education Details \r\n MCA   YMCAUST,  Faridab...
..            ...                                                ...
957       Testing  Computer Skills: • Proficient in MS office (...
958       Testing  ❖ Willingness to accept the challenges. ❖ ...
959       Testing  PERSONAL SKILLS • Quick learner, • Eagerne...
960       Testing  COMPUTER SKILLS & SOFTWARE KNOWLEDGE MS-Power ...
961       Testing  Skill Set OS Windows XP/7/8/8.1/10 Database MY...

[796 rows x 2 columns]

Let's see one example of a duplicated resume.

In [142]:
df_r.iloc[10]['Resume']

'Skills * Programming Languages: Python (pandas, numpy, scipy, scikit-learn, matplotlib), Sql, Java, JavaScript/JQuery. * Machine learning: Regression, SVM, Naïve Bayes, KNN, Random Forest, Decision Trees, Boosting techniques, Cluster Analysis, Word Embedding, Sentiment Analysis, Natural Language processing, Dimensionality reduction, Topic Modelling (LDA, NMF), PCA & Neural Nets. * Database Visualizations: Mysql, SqlServer, Cassandra, Hbase, ElasticSearch D3.js, DC.js, Plotly, kibana, matplotlib, ggplot, Tableau. * Others: Regular Expression, HTML, CSS, Angular 6, Logstash, Kafka, Python Flask, Git, Docker, computer vision - Open CV and understanding of Deep learning.Education Details \r\n\r\nData Science Assurance Associate \r\n\r\nData Science Assurance Associate - Ernst & Young LLP\r\nSkill Details \r\nJAVASCRIPT- Exprience - 24 months\r\njQuery- Exprience - 24 months\r\nPython- Exprience - 24 monthsCompany Details \r\ncompany - Ernst & Young LLP\r\ndescription - Fraud Investigatio

In [143]:
df_r[df_r['Resume'] == 'Skills * Programming Languages: Python (pandas, numpy, scipy, scikit-learn, matplotlib), Sql, Java, JavaScript/JQuery. * Machine learning: Regression, SVM, Naïve Bayes, KNN, Random Forest, Decision Trees, Boosting techniques, Cluster Analysis, Word Embedding, Sentiment Analysis, Natural Language processing, Dimensionality reduction, Topic Modelling (LDA, NMF), PCA & Neural Nets. * Database Visualizations: Mysql, SqlServer, Cassandra, Hbase, ElasticSearch D3.js, DC.js, Plotly, kibana, matplotlib, ggplot, Tableau. * Others: Regular Expression, HTML, CSS, Angular 6, Logstash, Kafka, Python Flask, Git, Docker, computer vision - Open CV and understanding of Deep learning.Education Details \r\n\r\nData Science Assurance Associate \r\n\r\nData Science Assurance Associate - Ernst & Young LLP\r\nSkill Details \r\nJAVASCRIPT- Exprience - 24 months\r\njQuery- Exprience - 24 months\r\nPython- Exprience - 24 monthsCompany Details \r\ncompany - Ernst & Young LLP\r\ndescription - Fraud Investigations and Dispute Services   Assurance\r\nTECHNOLOGY ASSISTED REVIEW\r\nTAR (Technology Assisted Review) assists in accelerating the review process and run analytics and generate reports.\r\n* Core member of a team helped in developing automated review platform tool from scratch for assisting E discovery domain, this tool implements predictive coding and topic modelling by automating reviews, resulting in reduced labor costs and time spent during the lawyers review.\r\n* Understand the end to end flow of the solution, doing research and development for classification models, predictive analysis and mining of the information present in text data. Worked on analyzing the outputs and precision monitoring for the entire tool.\r\n* TAR assists in predictive coding, topic modelling from the evidence by following EY standards. Developed the classifier models in order to identify "red flags" and fraud-related issues.\r\n\r\nTools & Technologies: Python, scikit-learn, tfidf, word2vec, doc2vec, cosine similarity, Naïve Bayes, LDA, NMF for topic modelling, Vader and text blob for sentiment analysis. Matplot lib, Tableau dashboard for reporting.\r\n\r\nMULTIPLE DATA SCIENCE AND ANALYTIC PROJECTS (USA CLIENTS)\r\nTEXT ANALYTICS - MOTOR VEHICLE CUSTOMER REVIEW DATA * Received customer feedback survey data for past one year. Performed sentiment (Positive, Negative & Neutral) and time series analysis on customer comments across all 4 categories.\r\n* Created heat map of terms by survey category based on frequency of words * Extracted Positive and Negative words across all the Survey categories and plotted Word cloud.\r\n* Created customized tableau dashboards for effective reporting and visualizations.\r\nCHATBOT * Developed a user friendly chatbot for one of our Products which handle simple questions about hours of operation, reservation options and so on.\r\n* This chat bot serves entire product related questions. Giving overview of tool via QA platform and also give recommendation responses so that user question to build chain of relevant answer.\r\n* This too has intelligence to build the pipeline of questions as per user requirement and asks the relevant /recommended questions.\r\n\r\nTools & Technologies: Python, Natural language processing, NLTK, spacy, topic modelling, Sentiment analysis, Word Embedding, scikit-learn, JavaScript/JQuery, SqlServer\r\n\r\nINFORMATION GOVERNANCE\r\nOrganizations to make informed decisions about all of the information they store. The integrated Information Governance portfolio synthesizes intelligence across unstructured data sources and facilitates action to ensure organizations are best positioned to counter information risk.\r\n* Scan data from multiple sources of formats and parse different file formats, extract Meta data information, push results for indexing elastic search and created customized, interactive dashboards using kibana.\r\n* Preforming ROT Analysis on the data which give information of data which helps identify content that is either Redundant, Outdated, or Trivial.\r\n* Preforming full-text search analysis on elastic search with predefined methods which can tag as (PII) personally identifiable information (social security numbers, addresses, names, etc.) which frequently targeted during cyber-attacks.\r\nTools & Technologies: Python, Flask, Elastic Search, Kibana\r\n\r\nFRAUD ANALYTIC PLATFORM\r\nFraud Analytics and investigative platform to review all red flag cases.\r\nâ\x80¢ FAP is a Fraud Analytics and investigative platform with inbuilt case manager and suite of Analytics for various ERP systems.\r\n* It can be used by clients to interrogate their Accounting systems for identifying the anomalies which can be indicators of fraud by running advanced analytics\r\nTools & Technologies: HTML, JavaScript, SqlServer, JQuery, CSS, Bootstrap, Node.js, D3.js, DC.js']

Category                                             Resume
0   Data Science  Skills * Programming Languages: Python (pandas...
10  Data Science  Skills * Programming Languages: Python (pandas...
20  Data Science  Skills * Programming Languages: Python (pandas...
30  Data Science  Skills * Programming Languages: Python (pandas...

Dropping duplicates:

In [144]:
df_r.drop_duplicates(subset = ['Resume'], keep = 'first', inplace = True)
df_r.reset_index(inplace = True, drop = True)

Number of resumes belonging to each of the job categories:

In [145]:
df_r["Category"].value_counts()

Category
Java Developer               13
Database                     11
Data Science                 10
Advocate                     10
HR                           10
DotNet Developer              7
Hadoop                        7
DevOps Engineer               7
Automation Testing            7
Testing                       7
Civil Engineer                6
Business Analyst              6
SAP Developer                 6
Health and fitness            6
Python Developer              6
Arts                          6
Electrical Engineering        5
Sales                         5
Network Security Engineer     5
Mechanical Engineer           5
ETL Developer                 5
Blockchain                    5
Operations Manager            4
Web Designing                 4
PMO                           3
Name: count, dtype: int64

Converting the column names to snake_case:

In [146]:
df_r.columns = df_r.columns.str.strip().str.lower()

In [148]:
df_r.to_csv('data/resumes.csv', index = False)